File: src/main/java/ink/rubi/coffee/controller/IController.java
package ink.rubi.coffee.controller;

import ink.rubi.coffee.po.Item;
import javafx.fxml.Initializable;
import javafx.scene.control.ChoiceBox;

/**
 * @author : Rubi
 * @version : 2019-05-17 23:41 下午
 */
public interface IController<T> extends Initializable {
    default T getConfigHolder() {
        return null;
    }

    default void flushConfig(T t) {
    }

    /**
     * for the communication between mainController and otherControllers.
     * @param mainController
     */
    void init(MainController mainController);


    default <I extends Item, T extends Enum<T>> void matchChoice(ChoiceBox<I> box, Class<T> enumClazz, String strValue, I defaultValue) {
        box.getSelectionModel().select(
                box.getItems()
                        .stream()
                        .filter(item -> item.getValue().equals(T.valueOf(enumClazz, strValue)))
                        .findAny()
                        .orElse(defaultValue));
    }
}

File: src/main/java/ink/rubi/coffee/controller/IController.kt
package ink.rubi.coffee.controller

import ink.rubi.coffee.po.Item
import javafx.fxml.Initializable
import javafx.scene.control.ChoiceBox
import java.util.*

/**
 * @author : Rubi
 * @version : 2019-05-17 23:41 下午
 */
interface IController<T> {
    fun getConfigHolder(): T

    fun flushConfig(holder: T) {}

    /**
     * for the communication between mainController and otherControllers.
     *
     * @param mainController
     */
    fun inject(mainController: MainController)

    fun defaultShow()

    fun <I : Item<*>, T : Enum<T>> matchChoice(box: ChoiceBox<I>, enumClazz: Class<T>, strValue: String, defaultValue: I) {
        box.selectionModel.select(
                box.items
                        .stream()
//                        .filter { item -> item.value == T.valueOf<T>(enumClazz, strValue) }
                        .filter { item -> item.value == EnumSet.allOf(enumClazz).find { t -> t.name == strValue } }
                        .findAny()
                        .orElse(defaultValue))
    }
}
